hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -914,6 +911,30 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
         return connection;
     }
 
+    private void checkClientStateOnClusterIdChange(TcpClientConnection connection) {
+        if (activeConnections.isEmpty()) {
+            //We only have single connection established
+            if (failoverConfigProvided) {
+                //If failover is provided, and this single connection is established after failover logic kicks in
+                // (checked via `switchingToNextCluster`), then it is OK to continue. Otherwise, we force the failover logic
+                // to be used by throwing `ClientNotAllowedInClusterException`
+                if (switchingToNextCluster) {
+                    switchingToNextCluster = false;
+                } else {
+                    String reason = "Force to hard cluster switch";
+                    connection.close(reason, null);
+                    throw new ClientNotAllowedInClusterException(reason);
+                }
+            }
+        } else {
+            //If there are other connections that means we have a connection to wrong cluster.
```

Comment:
What if there are other connections, but they're not yet closed by the `HeartbeatManager` (consider a situation when a cluster restarted fast enough)? Is it ok to close the connection in this case?

Reply:
Not sure if I understand the scenario correctly.  Let me try:
1. There were 2 members. 
2. The client is connected to the first one. 
3. While the client is trying to open the second connection, both members are restarted.
4. In this case we will close the connection to the second member first thinking that it is not part of the cluster we think we are in. 
5. The connection to the first cluster will be gone after that and we will initiate a reconnect to the cluster. 

Yes, you are right. It seems unfortunate that we close the connection to the second member at step 4. 
But there seems to be no way around this. It does not make sense to check if the connections on the active connections are closed or not. Because they can be closed right after we check them. 

Do you have a suggestion to avoid that?

Reply:
Yes, that scenario describes the problem. It could be also slightly different, but with the same outcome:

1. There were 2 members.
2. The client is connected to both members.
3. Both members are restarted in slightly different time and `HeartbeatManager` detects that for the connection to the first one. Then the re-connect logic kicks in.
4. In this case we will close the connection to the first member thinking that it is not part of the cluster we think we are in.
5. After a while we're going to start re-connect procedure for the second member.

Do you find this scenario possible?

As for the ways to avoid it, both scenarios may be not that critical critical, since we're going to keep trying to re-connect to all members in `ConnectToAllClusterMembersTask`. So, for the very last re-connection attempt (i.e. when `activeConnections.isEmpty() == true`) we should be able to connect to the new cluster. WDYT?

Reply:
After further discussion we came to a conclusion that both described scenarios shouldn't be a problem as the client should recover and connect to all members eventually.